src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs

```diff
@@ -0,0 +1,180 @@
+
+            _disposables.AddDisposable(dataSourceLink);
+            _disposables.AddDisposable(join);
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<string>>(nameFormat: nameof(DesignTimeInputsFileWatcher) + "Broadcast {1}");
+            _publicBlock = _broadcastBlock.SafePublicize();
+        }
+
+        private async Task ProcessDesignTimeInputs(IProjectVersionedValue<DesignTimeInputs> input)
+        {
+            // We don't want to watch any new files while we're disposing
+            if (IsDisposing)
+            {
+                return;
+            }
+
+            DesignTimeInputs designTimeInputs = input.Value;
+
+            IVsAsyncFileChangeEx vsAsyncFileChangeEx = await _fileChangeService.GetValueAsync();
+
+            // we don't care about the difference between types of inputs, so we just construct one hashset for fast comparisons later
+            var allFiles = new HashSet<string>(StringComparers.Paths);
+            allFiles.AddRange(designTimeInputs.Inputs);
+            allFiles.AddRange(designTimeInputs.SharedInputs);
+
+            // Remove any files we're watching that we don't care about any more
+            foreach (string file in _fileWatcherCookies.Keys.ToArray())
+            {
+                if (!allFiles.Contains(file))
+                {
+                    ImmutableInterlocked.TryRemove(ref _fileWatcherCookies, file, out uint cookie);
+                    await vsAsyncFileChangeEx.UnadviseFileChangeAsync(cookie);
+                }
+            }
+
+            // Now watch and output files that are new
+            foreach (string file in allFiles)
+            {
+                if (!_fileWatcherCookies.ContainsKey(file))
+                {
+                    // We don't care about delete and add here, as they come through data flow, plus they are really bouncy - every file change is a Time, Del and Add event)
+                    uint cookie = await vsAsyncFileChangeEx.AdviseFileChangeAsync(file, _VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size, sink: this);
+
+                    ImmutableInterlocked.TryAdd(ref _fileWatcherCookies, file, cookie);
+
+                    // Advise of an addition now
+                    PostToOutput(file);
+                }
+            }
+        }
+
+        private void PostToOutput(string file)
+        {
+            _version++;
+            ImmutableDictionary<NamedIdentity, IComparable> dataSources = ImmutableDictionary<NamedIdentity, IComparable>.Empty.Add(DataSourceKey, DataSourceVersion);
+
+            _broadcastBlock.Post(new ProjectVersionedValue<string>(file, dataSources));
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _broadcastBlock?.Complete();
+                _disposables.Dispose();
+
+                _threadingService.ExecuteSynchronously(async () =>
```

Comment:
This can overlap with ProcessDesignTimeInputs you need to protect both methods with a lock.

Reply:
Thanks. Going to use a CancellationTokenSource to cancel any processing that is happening when Dispose is called.

Reply:
Actually, that would still leave a small window for leaking a token. A semaphore probably works better, though is more likely to result in watching files, then unwatching them immediately.